micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,254 @@
+    }
+}
+
+enum PixelFormat {
+    /// 2BPP
+    RGB565 = 0,
+    /// 2BPP
+    YUV422 = 1,
+    /// 1.5BPP
+    YUV420 = 2,
+    /// 1BPP
+    GRAYSCALE = 3,
+    /// JPEG/Compressed
+    JPEG = 4,
+    /// 3BPP
+    RGB888 = 5,
+    RAW = 6,
+    /// 3BP2P
+    RGB444 = 7,
+    /// 3BP2P
+    RGB555 = 8,
+}
 
+/// Sizeof output image: QVGA|CIF|VGA|SVGA|XGA|SXGA|UXGA
+/// as u32
+enum FrameSize {
+    /// 96x96
+    W96XH96 = 0,
+    /// 160x120
+    QQVGA = 1,
+    /// 176x144
+    QCIF = 2,
+    /// 240x176
+    HQVGA = 3,
+    /// 240x240
+    W240XH240 = 4,
+    /// 320x240
+    QVGA = 5,
+    /// 400x296
+    CIF = 6,
+    /// 480x320
+    HVGA = 7,
+    /// 640x480
+    VGA = 8,
+}
+
+#[derive(DoCommand)]
 pub struct Esp32Camera {
     config: camera_config_t,
-    last_grab: Duration,
 }
 
 impl Esp32Camera {
-    pub fn new() -> Self {
-        let t = EspSystemTime;
-        Esp32Camera {
+    pub(crate) fn from_config(
+        cfg: ConfigType,
+        _: Vec<Dependency>,
+    ) -> Result<CameraType, CameraError> {
+        let pin_pwdn = cfg.get_attribute::<i32>("pin_pwdn").unwrap_or(-1);
+        let pin_reset = cfg.get_attribute::<i32>("pin_reset").unwrap_or(-1);
+        let pin_xclk = cfg.get_attribute::<i32>("pin_xclk").unwrap_or(21);
+        let pin_sccb_sda = cfg.get_attribute::<i32>("pin_sccb_sda").unwrap_or(26);
+        let pin_sccb_scl = cfg.get_attribute::<i32>("pin_sccb_scl").unwrap_or(27);
+        let pin_d7 = cfg.get_attribute::<i32>("pin_d7").unwrap_or(35);
+        let pin_d6 = cfg.get_attribute::<i32>("pin_d6").unwrap_or(34);
+        let pin_d5 = cfg.get_attribute::<i32>("pin_d5").unwrap_or(39);
+        let pin_d4 = cfg.get_attribute::<i32>("pin_d4").unwrap_or(36);
+        let pin_d3 = cfg.get_attribute::<i32>("pin_d3").unwrap_or(19);
+        let pin_d2 = cfg.get_attribute::<i32>("pin_d2").unwrap_or(18);
+        let pin_d1 = cfg.get_attribute::<i32>("pin_d1").unwrap_or(5);
+        let pin_d0 = cfg.get_attribute::<i32>("pin_d0").unwrap_or(4);
+        let pin_vsync = cfg.get_attribute::<i32>("pin_vsync").unwrap_or(25);
+        let pin_href = cfg.get_attribute::<i32>("pin_href").unwrap_or(23);
+        let pin_pclk = cfg.get_attribute::<i32>("pin_pclk").unwrap_or(22);
+        let xclk_freq_hz = cfg
+            .get_attribute::<i32>("xclk_freq_hz")
+            .unwrap_or(20_000_000);
+        let ledc_timer = cfg.get_attribute::<u32>("ledc_timer").unwrap_or(1);
+        let ledc_channel = cfg.get_attribute::<u32>("ledc_channel").unwrap_or(1);
+        let frame_size = cfg
+            .get_attribute::<u32>("frame_size")
+            .unwrap_or(FrameSize::QVGA as u32);
+        // Quality of JPEG output: 0-63 lower means higher quality
+        let jpeg_quality = cfg.get_attribute::<i32>("jpeg_quality").unwrap_or(32);
+        //  If pin_sccb_sda is -1, use the already configured I2C bus by number
+        let sccb_i2c_port = cfg.get_attribute::<i32>("sccb_i2c_port").unwrap_or(0);
```

Comment:
does this require that you've configured the i2c bus on board? If so, I think I'd actually like to see some validation in this regard (maybe by requiring an i2c bus name in the config and failing if we can't fetch the handle by name).

Reply:
Discussed offline with @gvaradarajan 
it appears that `esp_camera_init` accepts two options for i2c, either:
a) setting `pin_sccb_scl` and `pin_sccb_sda`, which it then initializes
b) choosing an already configured `sccb_i2c_port`, which it then attaches to.

when using the [m5 Camera Timer X](https://docs.m5stack.com/en/unit/timercam_x)'s ov3660 camera and values in `pin_sccb_sda`/`pin_sccb_scl`, I needed to set `sccb_i2c_port` to `-1` to prevent a low-level panic during `esp_cam_init`. 

I haven't confirmed what the behavior is if we use an i2c bus that we set in app and use those values here. When there are `i2c` pins set on the board, we unsafely initialize the pins with `unsafe AnyIOPin::new(pin)`. 

Follow up ticket: [RSDK-8327 ](https://viam.atlassian.net/browse/RSDK-8327)